pkg/services/ngalert/api/authorization.go

```diff
@@ -162,3 +171,90 @@ func (api *API) authorize(method, path string) web.Handler {
+	}
+	return ac.EvalAll(scopes...)
+}
+
+// authorizeRuleChanges analyzes changes in the rule group, determines what actions the user is trying to perform and check whether those actions are authorized.
+// If the user is not authorized to perform the changes the function returns ErrAuthorization with description of what action is not authorized. If evaluator function returns error, the function returns it
+func authorizeRuleChanges(namespace *models.Folder, changes *changes, evaluator func(evaluator ac.Evaluator) (bool, error)) error {
+	namespaceScope := dashboards.ScopeFoldersProvider.GetResourceScope(strconv.FormatInt(namespace.Id, 10))
+	if len(changes.Delete) > 0 {
+		allowed, err := evaluator(ac.EvalPermission(ac.ActionAlertingRuleDelete, namespaceScope))
+		if err != nil {
+			return fmt.Errorf("failed to perform user authorization: %w", err)
+		}
+		if !allowed {
+			return fmt.Errorf("%w user cannot delete alert rules that belong to folder %s", ErrAuthorization, namespace.Title)
+		}
+	}
+
+	// check that the user can create alerts in
+	if len(changes.New) > 0 {
+		allowed, err := evaluator(ac.EvalPermission(ac.ActionAlertingRuleCreate, namespaceScope))
+		if err != nil {
+			return fmt.Errorf("failed to perform user authorization: %w", err)
+		}
+		if !allowed {
+			return fmt.Errorf("%w user cannot create alert rules in the folder %s", ErrAuthorization, namespace.Title)
+		}
+	}
+
+	updateChecked := false
+
+	for _, rule := range changes.Update {
+		if rule.Existing.NamespaceUID != rule.New.NamespaceUID {
+			allowed, err := evaluator(ac.EvalAll(ac.EvalPermission(ac.ActionAlertingRuleDelete, dashboards.ScopeFoldersProvider.GetResourceScopeUID(rule.Existing.NamespaceUID))))
+			if err != nil {
+				return fmt.Errorf("failed to perform user authorization: %w", err)
+			}
+			if !allowed {
+				return fmt.Errorf("%w to delete alert rules from folder UID %s", ErrAuthorization, rule.Existing.NamespaceUID)
+			}
+			allowed, err = evaluator(ac.EvalPermission(ac.ActionAlertingRuleCreate, namespaceScope))
+			if err != nil {
+				return fmt.Errorf("failed to perform user authorization: %w", err)
+			}
+			if !allowed {
+				return fmt.Errorf("%w to create alert rules in the folder '%s'", ErrAuthorization, namespace.Title)
+			}
+		} else if !updateChecked {
+			allowed, err := evaluator(ac.EvalAll(ac.EvalPermission(ac.ActionAlertingRuleUpdate, namespaceScope)))
+			if err != nil {
+				return fmt.Errorf("failed to perform user authorization: %w", err)
+			}
+			if !allowed {
+				return fmt.Errorf("%w to update alert rules that belong to folder %s", ErrAuthorization, namespace.Title)
+			}
+			updateChecked = true
+		}
+
+		// if data were not changed, skip authorization. Effectively, this will let users without read access for
+		// data sources limited set of settings, like interval, for, labels, annotations.
+		if len(rule.Diff.GetDiffsForField("Data")) > 0 {
```

Comment:
This has the potential to be dangerous, what you're saying here is _everything that's not data_ you can access. However, if we at a later date add a field that's meant to be editable under this scenario, we would.

Can we be explicit with our allow list?

Reply:
I will clarify the comment. 
What it actually means that in order to change the rule's queries (as well as expressions) the user should have permission `datasource:query` for all data sources that mentioned in the rule. That is on top of the `alert.rules:update` permissions. 

Therefore, if user has permission `alert.rules:update` but does not have permission `datasource:query` for one or many datasources used by the rule, the user can change title, interval, for, labels, annotations, condition etc.

The idea is to let users to change some aspects of the rule without having access to the data source.